source/Ship.cpp

```diff
@@ -2003,6 +2003,20 @@ void Ship::DoGeneration()
 	// maximum capacity for the rest of the turn, but must be clamped to the
 	// maximum here before they gain more. This is so that, for example, a ship
 	// with no batteries but a good generator can still move.
+
+	// Run fuel consumption in reverse to get rid of excess energy from the last frame
+	// so that energy CAN be available for the frame, but if it isn't used, then the
+	// fuel consumption retroactively didn't happen, and we just don't talk about it.
+	// This also has the effect of shutting off fuel consumption completely if another
+	// energy-generating alternative can cover energy usage for the previous frame.
+	if(attributes.Get("fuel consumption") && energy > attributes.Get("energy capacity"))
+	{
+		double scale = min(1., (energy - attributes.Get("energy capacity")) / attributes.Get("fuel energy"));
+		energy -= scale * attributes.Get("fuel energy");
+		heat -= scale * attributes.Get("fuel heat");
+		fuel += scale * attributes.Get("fuel consumption");
+	}
+
```

Comment:
You should still be able to make use of DoRepair for this, just with switching around some of the values.

```c++
DoRepair(fuel, fuelConsumption, attributes.Get("fuel capacity"), energy, fuelEnergy, fuel, 0., heat, -fuelHeat);
```

Given that we'd be making use of these attributes in two spots then (once for energy to fuel then a second time for fuel to energy), you'd want to move the variables from lower down (L2078-2080 right now) up to here and make them const.
```c++
const double fuelEnergy = attributes.Get("fuel energy");
const double fuelConsumption = attributes.Get("fuel consumption") / fuelEnergy;
const double fuelHeat = attributes.Get("fuel heat") / fuelEnergy;
```

Also, a comment as simple as `// Convert any excess energy back into fuel if able to.` would do.

But I think that this is an entirely separate issue that should be handled elsewhere, as while I can get on board with fuel consumption and generation being bounded by energy and heat costs just as shield and hull repair are bounded, this whole "convert excess energy back into fuel" thing is a new mechanic entirely.